src/main/java/com/google/api/codegen/configgen/CollectionPattern.java

```diff
@@ -34,45 +38,136 @@ public static CollectionPattern create(FieldSegment fieldSegment) {
+   * @param wildcardIndex the index of the wildcard segment
+   * @return the entity name that corresponds to the wildcard segment
+   */
+  private static String buildEntityName(List<PathSegment> pathSegments, int wildcardIndex) {
+    String prefix = buildEntityPrefix(pathSegments, wildcardIndex);
+    return buildEntityName(pathSegments, wildcardIndex, prefix);
+  }
+
+  /**
+   * Builds an entity name for a wildcard path segment.
+   *
+   * @param pathSegments the full path containing the wildcard segment at wildcardIndex
+   * @param wildcardIndex the index of the wildcard segment
+   * @param prefix the prefix to the entity name. Should be the resource name.
+   * @return the entity name that corresponds to the wildcard segment
+   */
+  private static String buildEntityName(
+      List<PathSegment> pathSegments, int wildcardIndex, String prefix) {
+    WildcardSegment wildcardSegment = (WildcardSegment) pathSegments.get(wildcardIndex);
+    return wildcardSegment.isUnbounded() ? String.format("%s_path", prefix) : prefix;
+  }
+
+  /**
+   * Builds the entity prefix for a wildcard path segment.
+   *
+   * <p>The entity prefix is just the entity name for the collection resource without handling
+   * singular resources or unbounded wildcards.
+   *
+   * @param pathSegments the full path containing the wildcard segment at wildcardIndex
+   * @param wildcardIndex the index of the wildcard segment
+   * @return the entity prefix that corresponds to the wildcard segment
+   */
+  private static String buildSimpleEntityPrefix(List<PathSegment> pathSegments, int wildcardIndex) {
+    PathSegment wildcardSegment = pathSegments.get(wildcardIndex);
+    if (wildcardIndex == 0) {
+      return "unknown";
     }
-    templatizedSubpath = builder.build();
 
-    if (simpleName == null) {
-      // This should never occur because of isValidCollectionPattern checks
-      throw new IllegalStateException("Field segment contained no wildcards.");
+    PathSegment prevSegment = pathSegments.get(wildcardIndex - 1);
+    if (!(prevSegment instanceof LiteralSegment)) {
+      return "unknown";
     }
+
+    return LanguageUtil.upperCamelToLowerUnderscore(Inflector.singularize(prevSegment.syntax()));
+  }
+
+  /**
+   * Builds the entity prefix for a wildcard path segment, handling singular resources.
+   *
+   * <p>The entity prefix for singular resources contains the entity name of the parent collection.
+   *
+   * @param pathSegments the full path containing the wildcard segment at wildcardIndex
+   * @param wildcardIndex the index of the wildcard segment
+   * @return the entity prefix that corresponds to the wildcard segment
+   */
+  private static String buildEntityPrefix(List<PathSegment> pathSegments, int wildcardIndex) {
+    String collectionEntityName = buildSimpleEntityPrefix(pathSegments, wildcardIndex);
+
+    // If there are an odd number of remaining segments, then the last segment is for a singular
```

Comment:
We only ever call this on the last wildcard in a template. So what about checking for `remainingSegments > 0` instead of for an even number of segments?

Reply:
Done